### PR TITLE
Automatic disc change for 2-disc games

### PR DESCRIPTION
**Pull request by JosJuice**
This adds a setting that makes Dolphin automatically switch discs when you reach a point in a 2-disc game where the game prompts you to switch discs. The setting is off by default.

In order for Dolphin to know which other disc to insert, the game has to be started in one of three ways:

* From the game list, with both discs being present in the game list. (This currently also works if you use the Default ISO feature and the other disc is in the game list, but I'm not sure if it should. Any opinions?)
* With File > Open or the command line interface, with the paths to both discs being provided.
* By launching an M3U file with File > Open or the command line interface. ~~(Not implemented yet)~~

If you start the game in another way, Dolphin just won't switch discs automatically.

Left to do:

- [x] M3U file support
- [x] Android game list support
- [x] Don't do anything funny while the Wii Menu is running
- [x] Probably add a delay to the automatic disc change (games can get confused if the disc gets ejected immediately)

### CODE REVIEW

**Comment by JosJuice** [comment]
Also, I replaced the old system of automatic disc switching for input recordings with this new system. The old system relied on there being a file with a specific name (determined by the DTM header) in the same folder as the first disc.

**Comment by JMC47** [comment]
Do we have a list of multi-disc games to test?  I can easily test MGS and Tales of Symphonia, but, there are others.

Does it work for the Dragon Quest X installer lol.

**Comment by tony971** [comment]
> Do we have a list of multi-disc games to test? I can easily test MGS and Tales of Symphonia, but, there are others.

* Baten Kaitos - Eternal Wings and the Lost Ocean
* Baten Kaitos Origins
* Enter the Matrix
* Freaky Flyers
* GoldenEye - Rogue Agent
* Killer 7
* Medal of Honor - Rising Sun
* Metal Gear Solid - The Twin Snakes
* Lord of the Rings, The - The Third Age
* Pokemon Colosseum (If you include the bonus disc)
* Resident Evil
* Resident Evil 4
* Resident Evil - Code - Veronica X
* Resident Evil Zero
* Ribbit King Plus! (If you include the bonus disc)
* Tales of Symphonia
* Teenage Mutant Ninja Turtles 2 - Battle Nexus
* Teenage Mutant Ninja Turtles 3 - Mutant Nightmare
* Tiger Woods PGA Tour 2004
* Tiger Woods PGA Tour 2005
* Tom Clancy's Splinter Cell - Chaos Theory
* Tom Clancy's Splinter Cell - Double Agent

Plus any PAL/Japanese exclusives.

**Comment by JosJuice** [comment]
> Does it work for the Dragon Quest X installer lol.

I have no idea what that one does, so I don't know if it'll work. Hopefully all GameCube games should work, at least after I add a delay to the disc switching.

@tony971 I don't think games with bonus discs count for this, as the feature only activates when the game prompts you to switch discs.

**Comment by JosJuice** [comment]
I've added a 500 ms delay. The game I've been testing, Baten Kaitos, only shows the prompt to switch discs and the "Reading disc..." text if there's a delay, though it does work other than that regardless of whether there is a delay.

**Comment by JosJuice** [comment]
When it comes to M3U files, would it be fine to only support UTF-8? The Wikipedia page for the M3U format says that .m3u uses the system encoding and .m3u8 uses UTF-8, but I suppose that only is based on what media players do, not emulators. Does anyone know how other emulators that use M3U files for multi-disc support handle encodings?

**Comment by kevinvanrijn** [comment]
The [Wikipedia page](https://en.wikipedia.org/wiki/M3U#M3U8) also says that the current proposal is to not distinguish between "m3u" and "m3u8".

Afaik, audio players are forced to read both as UTF-8 for things to "just work".

**Comment by JosJuice** [comment]
That's the current proposal for the HLS format, not for M3U files in general. But I suppose it does set some kind of precedent.

**Comment by JosJuice** [comment]
This PR should be feature complete now. I decided to only support UTF-8 in M3U files.

File: Source/Core/Core/Boot/Boot.cpp [hunk 1]
```diff
@@ -54,47 +60,122 @@
...
 {
 }
 
 std::unique_ptr<BootParameters>
-BootParameters::GenerateFromFile(const std::string& path,
+BootParameters::GenerateFromFile(std::string boot_path,
                                  const std::optional<std::string>& savestate_path)
 {
-  const bool is_drive = Common::IsCDROMDevice(path);
+  return GenerateFromFile(std::vector<std::string>(), savestate_path);
```
Comment (riking) [inline Source/Core/Core/Boot/Boot.cpp]: vector needs to include boot_path, this will just always trip the ASSERT

Comment (JosJuice) [reply Source/Core/Core/Boot/Boot.cpp]: Oops. Thanks for catching that.

**Comment by JMC47** [comment]
Double checked this on Android now to make sure it was working.  Seems to be good.

**Comment by JMC47** [comment]
Perhaps slightly late, but this appears to work on netplay/movies as well.

**Pull request merged**